### ISSUE TITLE
[CLOSED] Custom Default Project Templates

### ISSUE THREAD

**Issue by schuhschuh** [cmake-basis/BASIS#248]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [schuhschuh](https://github.com/schuhschuh)**
_Thursday Nov 14, 2013 at 23:51 GMT_
_Originally opened as https://github.com/schuhschuh/cmake-basis-legacy/issues/248_

---

Add CMake variable `DEFAULT_TEMPLATE` which can be set to a custom project template. If such custom template is specified, use and install it instead of the one shipped with BASIS itself. Moreover, if `INSTALL_TEMPLATE_DIR` is set to an empty string, do not install the template at all. This is useful when the custom project template lives in specific location accessible by the BASIS installation and will remain there. Then it is not necessary to copy the files into the BASIS installation.

This behavior makes #247 obsolete. The boolean switch is implicit to whether or not a custom template was specified.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Thursday Nov 14, 2013 at 23:54 GMT_

---

Andrew, can you please test these changes with your custom template ? If it works fine, I will merge it into the develop branch.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Friday Nov 15, 2013 at 21:08 GMT_

---

Continuing conversation from where I accidentally posted in #226 

I found and fixed the space, but I ran into some additional unexpected behavior, where many of the files weren't copied from the template, such as the CMakeLists.txt files.

I also have a directory:

```
/BASISTemplate
     /BASISTemplate-1.0      
```

Should I point TEMPLATE_DIR to BASISTemplate or BASISTemplate-1.0?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Friday Nov 15, 2013 at 21:10 GMT_

---

Either, it should not matter. See modifications of `config/Settings.cmake` and `data/CMakeLists.txt` files.

Edit: I missed btw your directory hierarchy. So it would be `BASSTemplate/BASISTemplate-1.0` or just `BASSTemplate/BASISTemplate`. But I guess you figured already.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Friday Nov 15, 2013 at 21:11 GMT_

---

I was thinking that TEMPLATE_DIR should be the higher level directory because it makes it easier to support multiple templates in one location. I expect we will add a second included template to BASIS anyway so there is a non-sbia version available. This would be a separate issue, but would it be worthwhile to move template-1.0 ... to sbiaTemplate/sbiaTemplate-1.0...?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Friday Nov 15, 2013 at 21:15 GMT_

---

Probably worthwhile, but I would use the naming scheme `/path/to/template/named/<name>/<version>/` probably instead or `/path/to/template/named/<name>/template-<version>`.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Friday Nov 15, 2013 at 21:20 GMT_

---

I'd suggest `/<name>/<version>`

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Friday Nov 15, 2013 at 21:23 GMT_

---

Yes, I think I like that better as well.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Friday Nov 15, 2013 at 21:27 GMT_

---

It also seems the path is set to `Installing: /opt/basis/share/template/`

Wit the current code, Is that where I would place `myTemplate/myTemplate-1.0/`? That would make the whole path:  `/opt/basis/share/template/myTemplate/myTemplate-1.0/`, then in the next version it would be `/opt/basis/share/template/myTemplate/1.0/`.

I'm also running into a problem with my template installer. I used a modified version of the one included in basis but it seems to skip installing some file types in a way I wouldn't expect, such as all CMakeLists.txt files. Do you have any idea what may be the problem?

```
##############################################################################
# @file  CMakeLists.txt
# @brief Build configuration of auxiliary data files.
#
# Copyright (c) 2011, 2012 University of Pennsylvania. All rights reserved.<br />
# See http://www.rad.upenn.edu/sbia/software/license.html or COPYING file.
#
# Copyright (c) 2013 Carnegie Mellon University
#
# Contact: SBIA Group <sbia-software at uphs.upenn.edu>
##############################################################################

macro (basis_install_template VERSION)
  install (
    DIRECTORY   "${CMAKE_CURRENT_SOURCE_DIR}/nrecBASISTemplate-${VERSION}/"
    DESTINATION "${INSTALL_TEMPLATE_DIR}/nrecBASISTemplate-${VERSION}"
    PATTERN     *~             EXCLUDE
    PATTERN     .svn           EXCLUDE
    PATTERN     .git           EXCLUDE
    PATTERN     .hg            EXCLUDE
    PATTERN     .DS_Store      EXCLUDE
  )
endmacro ()

# Attention: Whenever a template file is modified or added,
#            a new template directory with an updated version
#            number (<major>.<minor>) has to be created and the
#            modified or added files saved under this new directory.
#            As only the "differences" are stored each time,
#            *all* template versions are important to keep !
#            Otherwise, the three-way diff merge used by the basisproject
#            tool to update existing projects will fail.

if (BUILD_PROJECT_TOOL)
  basis_install_template (1.0)
endif ()

```

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Friday Nov 15, 2013 at 21:28 GMT_

---

My modification actually will install the template to `INSTALL_TEMPLATE_DIR` if not set to empty. You don't have to take care of this. It will copy all versions. See `data/CMakeLists.txt` of BASIS.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Friday Nov 15, 2013 at 21:39 GMT_

---

Ah well then your script is the one skipping my CMakeLists.txt files. :-)

The system also seems like it still assumes that the template version will be 1.8. I run the following series of commands:

```
rm -rf /opt/basis/*; 
rm -rf ../buildbasis/*; 
rm -rf ../testbasis/*; 
cmake ../basis -DTEMPLATE_DIR=~/Source_Code/coretech/basis/nrecBASISTemplate/ -DTEMPLATE_VERSION=1.0; 
make -j8 install; 
cd ../testbasis; 
/opt/basis/bin/basisproject --responsefile ~/Source_Code/coretech/basis/nrecBASIS.rsp; 
cd ../buildbasis
```

And it builds/installs, but skips some CMakeLists.txt files and can't find the template. I the end of the script prints the following output:

```
-- Installing: /opt/basis/share/template/BASISQuickStart.sh
-- Installing: /opt/basis/share/template/CMakeLists.skip.txt
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/_config.py
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/AUTHORS.txt
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/BasisProject.cmake
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/ChangeLog.txt
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/Components.cmake
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/Config.cmake.in
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/ConfigSettings.cmake
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/ConfigUse.cmake.in
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/ConfigVersion.cmake.in
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/CTestCustom.cmake.in
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/Package.cmake
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/ScriptConfig.cmake.in
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/config/Settings.cmake
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/COPYING.txt
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/CTestConfig.cmake
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/changelog.rst
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/download.rst
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/doxygen_extra.css.in
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/doxygen_footer.html.in
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/doxygen_header.html.in
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/index.rst
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/installation.rst
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/manual.rst
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/nrec_logo.png
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/people.rst
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/doc/publications.rst
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/INSTALL.txt
-- Installing: /opt/basis/share/template/nrecBASISTemplate-1.0/README.txt
-- Installing: /opt/basis/share/template/README.txt
-- Installing: /opt/basis/lib/cmake/basis/BASISUninstall.cmake
-- Installing: /opt/basis/bin/uninstall-
-- Installing: /opt/basis/lib/cmake/basis/BASISInstallManifest.txt
Missing template configuration file: /opt/basis/share/template-1.8/_config.py
Please specify a valid project template directory using --template.
```

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Friday Nov 15, 2013 at 21:46 GMT_

---

Ok, the template version is indeed still the same. You cannot overwrite it. I'll have a look.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Friday Nov 15, 2013 at 21:48 GMT_

---

I pushed commit b9ca251 which merged develop into this branch, since some of the changes were useful for testing. That way we can test with the same changeset, I know it breaks purity a bit...

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Saturday Nov 16, 2013 at 01:26 GMT_

---

Improved upon the initial attempt and changed it all over. The relevant option is now called `DEFAULT_TEMPLATE` which can be either set to `<template>/<version>`, where `<template>` is the name of a template provided by BASIS (e.g., `sbia`) or a full path to a custom template such as `<path>/<template>/<version>`. If a relative path is specified, the binary directory from which CMake was executed is used to make it absolute. The default template is installed to `INSTALL_TEMPLATE_DIR` if `BUILD_PROJECT_TOOL` is `ON` and a valid destination is set through the advanced `INSTALL_TEMPLATE_DIR` variable. It is by default `share/templates`. The (additional) templates provided by BASIS itself are installed to the same directory if the option `INSTALL_TEMPLATES` is set to `ON`.

See commit c068ac8beacc5859d7255b62b046df77d2c5914d.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Saturday Nov 16, 2013 at 19:55 GMT_

---

Sounds great, I think your comment here would also be useful if it were added directly to the documentation somewhere.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 18, 2013 at 20:26 GMT_

---

I need to have some empty defaults where the user can still supply a new default to basisproject, and right now I have the following:

_config.py:

``` python
  'division-name' : {
    'help'    : "Name of the package provider division or group.",
    'default' : ""
  },
  'division-website' : {
    'help'    : "Website of the package provider division or group.",
    'default' : ""
  },
```

BasisProject.cmake template:

```
DIVISION_NAME <division-name>
DIVISION_LOGO <division-website>
```

BasisProject.cmake output:

```
DIVISION_NAME ""
DIVISION_LOGO ""
```

Will this be interpreted correctly as not defined, or will it attempt to render the empty string image + division and consequently not work properly?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Monday Nov 18, 2013 at 20:31 GMT_

---

If the type of the argument is `string`, it will be enclosed by double quotes as part of the substitution

``` python
if type(value) is str: value = '"' + value.replace('"', '\\"') + '"'
```

and any double quotes within the string are escaped by a backslash.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 18, 2013 at 20:35 GMT_

---

Whoops, I phrased my question incorrectly. What I'm actually asking is how will `basis_project()` interpret `DIVISION_LOGO ""`, Will it break things or correctly interpret that there is no logo?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Monday Nov 18, 2013 at 21:03 GMT_

---

The CMake function used for the parsing will do what you expect and set `PROJECT_DIVISION_LOGO` to an empty string. When used in an `if` condition such as

``` cmake
if (PROJECT_DIVISION_LOGO)
```

this block will not be executed. Your safe. I usually make just like CMake no difference between an undefined variable or one set to an empty string. That is pretty much the same as long as you don't use the `if (DEFINED)` command, which I generally don't use.

It would be problematic to not have the double quotes for the empty string because then the CMake parsing function might likely just take the name of the next option, e.g., `DIVISION_WEBSITE`, as value of `DIVISION_LOGO`.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 18, 2013 at 21:49 GMT_

---

Great, it seems to have worked. It seems like there might still be an issue for the package logo.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Thursday Nov 21, 2013 at 23:10 GMT_

---

It seems simply using "LOGO path/to/logo.png" does not work, but "PACKAGE_LOGO path/to/logo.png" does.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Friday Nov 22, 2013 at 00:36 GMT_

---

The option `LOGO` of the `basis_project` command does not exist. It is indeed named `PACKAGE_LOGO`.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Friday Nov 22, 2013 at 18:34 GMT_

---

Ah, in some of the code LOGO is noted as an alias for PACKAGE_LOGO. I'll find and delete that. Once that is done I think the only thing remaining for this issue is documentation.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Friday Nov 22, 2013 at 22:02 GMT_

---

I will resolve the `PACKAGE_LOGO` problem as part of issue #252.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Monday Nov 25, 2013 at 20:27 GMT_

---

Sorry, my mistake. As you noted that `LOGO` did not work for you, I just thought I had not added it without checking the implementation again. But actually I wanted to allow it as short version for `PACKAGE_LOGO`.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Tuesday Dec 24, 2013 at 02:42 GMT_

---

The DEFAULT_TEMPLATE option is documented as part of the BASIS installation instructions. The custom template creation is partially documented as part of a respective how-to guide. Marking this issue as completed.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Tuesday Dec 24, 2013 at 03:02 GMT_

---

do we have commit numbers for those somewhere?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Tuesday Dec 24, 2013 at 03:19 GMT_

---

Do you remember when you added the extra section documenting the BASIS installation options ? That's when you added it first.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Tuesday Dec 24, 2013 at 03:24 GMT_

---

First commit related to this was probably 7318a70827cf7dfe73c21e2bc686d62802515571, but this has changed quite some since then.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Tuesday Dec 24, 2013 at 03:24 GMT_

---

The more recent commit is 5b7edabdcf360843a76bc690dae948eea4ffde85 and then 27781f291161f339a9677f084e69ca238830a44a.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Tuesday Dec 24, 2013 at 05:53 GMT_

---

Thanks for getting those.
